main/src/ca/uwaterloo/flix/language/phase/Reducer.scala

```diff
@@ -33,131 +33,152 @@ import scala.jdk.CollectionConverters._
   * 2. Collect a set of all anonymous class / new object expressions
   * 3. Collect a flat set of all types of the program, i.e., if `List[String]` is
   *   in the list, so is `String`.
-  *
-  * This is actually achieved as two pseudo phases, first 1. and 2. and then 3.
-  * operating on a new root.
   */
 object Reducer {
 
   def run(root: Root)(implicit flix: Flix): Root = flix.phase("Reducer") {
-    implicit val ctx: SharedContext = SharedContext(new ConcurrentLinkedQueue)
+    implicit val ctx: SharedContext = SharedContext(new ConcurrentLinkedQueue, new ConcurrentLinkedQueue)
 
     val newDefs = ParOps.parMapValues(root.defs)(visitDef)
+    val defTypes = ctx.defTypes.asScala.toSet
 
-    val newRoot = root.copy(defs = newDefs, anonClasses = ctx.anonClasses.asScala.toList)
-    val types = typesOf(newRoot)
-    newRoot.copy(types = types)
+    val effectTypes = root.effects.values.toSet.flatMap(typesOfEffect)
+
+    val types = nestedTypesOf(Set.empty, Queue.from(defTypes ++ effectTypes))
```

Comment:
Since this is single threaded, can you try and make the SharedContext hold a concurrent set of types instead? just so there's no duplicates to go over. I think it should be faster but you can run benchmarks for both

Reply:
OK, can we do that? Or was it already done?

Reply:
It's done